### PR TITLE
Merge the context manager and decorator

### PR DESCRIPTION
**Pull request by j00bar**
Instead of having different objects for setting logging context with a decorator vs. the context manager, this PR combines the two and makes them one.

This also removes the asymmetry in calling them -- instead of the decorator using kwargs as its context setting and the context manager using a passed dictionary, it uses kwargs for both.

Additionally, the LoggerContext `__init__` arguments are renamed to have underscore prefixes so as to reduce the likelihood of collision with a configured context variable.

Also, re-enables some testing for the overall integrated system that were `@skip`'ed before.

### CODE REVIEW

File: woodchipper/context.py [hunk 1]
```diff
@@ -1,16 +1,81 @@
...
+
+        return_mapping[dig_path_head] = ParamSearchConfig(
+            dig_path=dig_path_tail,
+            logvar_name=logvar_name,
+            diggable=True if dig_path_tail else False,
+        )
+
+    return return_mapping
+
+
```
Comment (j00bar) [inline woodchipper/context.py:78]: This is just moved from `decorators.py`

File: woodchipper/context.py [hunk 2]
```diff
@@ -1,16 +1,81 @@
...
+    supports mapping-like behavior so cannot search sequences."""
+    if not str_path:
+        return obj
+    else:
+        head, remaining = _split_head_node(str_path, delimiter)
+        value = _get_key_or_attr(obj, head)
+        if value is missing:
+            return missing
+        else:
+            return pluck_value(value, remaining, delimiter)
```
Comment (gtwalford) [inline woodchipper/context.py:53]: 👍 very clever!

File: woodchipper/context.py [hunk 3]
```diff
@@ -94,19 +175,47 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    def __call__(self, f):
+        self.decorator_mapping = _build_path_head_to_param_config_map(
+            self.injected_context, delimiter=self.path_delimiter
+        )
 
-logging_ctx = LoggingContextVar("logging_ctx")
+        @wraps(f)
+        def wrapper(*func_args, **func_kwargs):
 
+            # This will pair up the function signature parameters with the arguments that were passed. It's ...amazing.
```
Comment (gtwalford) [inline woodchipper/context.py:186]: 😲

File: woodchipper/context.py [hunk 4]
```diff
@@ -94,19 +175,47 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
-            return str(value)
-        except Exception:
-            raise ValueError(
-                f"Value type (type={type(value)} is not a not a LoggableType ({str, int, bool, Decimal, float, None}) "
-                f"and cannot be converted to string."
-            )
+            # Note: We not will dig into the *args or **kwargs of the function, if the function definition used
+            # *args or **kwargs. It's more difficult for **kwargs and not possible with *args.
+            # TODO: attempt digging into function **kwargs as well?
+            # TODO: consider a configurable 'log_defaults' decorator parameter that will allow the logging of
```
Comment (gtwalford) [inline woodchipper/context.py:193]: Should these be added as enhancements in the issues for the project?

File: woodchipper/context.py [hunk 5]
```diff
@@ -94,19 +175,47 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            for dec_param_name, param_config_entry in self.decorator_mapping.items():
+                if dec_param_name not in mapped_args:
+                    raw_value = self.missing_default
+                else:
+                    if param_config_entry.diggable:
+                        raw_value = pluck_value(
+                            mapped_args[dec_param_name],
+                            param_config_entry.dig_path,
+                            delimiter=self.path_delimiter,
+                        )
```
Comment (gtwalford) [inline woodchipper/context.py:209]: This is very cool - took me a minute to grok it all but the demo examples help

**Review by gtwalford — APPROVED** [review]
Very cool! I like the dot notation handling within the decorator. Should definitely get this tested to ensure all cases with the pluck_values.

File: woodchipper/context.py (hunk 4 above)
Comment (j00bar) [reply woodchipper/context.py:193]: @Quidge What do you think?

Comment (Quidge) [reply woodchipper/context.py:193]: Yeah I definitely think a flag to log the defaults would be valuable.

I think there is less demand for the capability to dig into function kwargs. If an argument is in the `**kwargs` collection, it's likely that the function being decorated isn't particularly concerned with whatever is in there (else it'd be explicitly named and not in `kwargs`). There are cases where this isn't true, but they're less common.

I think it'd be worth an enhancement flag but it's not a high priority imo.

**Pull request merged**